### PR TITLE
Add importlib.metadata (Python >= 3.8)

### PR DESCRIPTION
**Pull request by srittau**
Part of #3319

### CODE REVIEW

**Comment by srittau** [comment]
@rchen152 I have another pytype failure that I am unclear about: https://travis-ci.org/python/typeshed/jobs/596595124. "TypeError: Unexpected definition type <class 'NoneType'>", although all uses of `None` look ok to me and not like something that hasn't been done before in other stubs. Maybe it's the `@overload` + `@classmethod`? Unfortunately I can't get pytype to run locally:

```
$ ~/.pyenv/versions/3.6.9/bin/pytype stdlib/3/importlib/metadata.pyi 
usage: pytype [options] input [input ...]
pytype: error: Need an input.
```

**Comment by rchen152** [comment]
I'll take a look. To run pytype locally, the command is `pytype-single --parse-pyi --module-name=<whatever> -V<major.minor> <filepath>`.

**Comment by rchen152** [comment]
Huh. pytype can't parse nested classes inside a version check.
```
class Foo:
  class Bar: ...
```
parses fine.
```
if sys.version_info >= (3, 8):
  class Foo:
    class Bar: ...
```
crashes the parser. I'll work on a fix.

**Comment by srittau** [comment]
This passes CI now

File: stdlib/3/importlib/metadata.pyi [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+import os
+import pathlib
+import sys
+from email.message import Message
+from importlib.abc import MetaPathFinder
+from typing import Any, Dict, Iterable, List, NamedTuple, Optional, Tuple, Union, overload
+
+if sys.version_info >= (3, 8):
+    class PackageNotFoundError(ModuleNotFoundError): ...
+    class EntryPoint(NamedTuple("EntryPointBase", [("name", str), ("value", str), ("group", str)])):
```
Comment (JelleZijlstra) [inline stdlib/3/importlib/metadata.pyi]: We can define these with variable annotations now (thanks for doing the work to make that happen).

File: stdlib/3/importlib/metadata.pyi [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+        dist: Distribution
+    class FileHash:
+        mode: str
+        value: str
+        def __init__(self, spec: str) -> None: ...
+    class Distribution:
+        @abc.abstractmethod
+        def read_text(self, filename: str) -> Optional[str]: ...
+        @abc.abstractmethod
+        def locate_file(self, path: os.PathLike[str]) -> os.PathLike[str]: ...
```
Comment (JelleZijlstra) [inline stdlib/3/importlib/metadata.pyi]: Shouldn't the argument accept str?

File: stdlib/3/importlib/metadata.pyi [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+        @property
+        def entry_points(self) -> List[EntryPoint]: ...
+        @property
+        def files(self) -> Optional[List[PackagePath]]: ...
+        @property
+        def requires(self) -> Optional[List[str]]: ...
+    class DistributionFinder(MetaPathFinder):
+        class Context:
+            name: Optional[str]
+            def __init__(self, *, name: Optional[str] = ..., path: List[str] = ..., **kwargs: Any) -> None: ...
```
Comment (JelleZijlstra) [inline stdlib/3/importlib/metadata.pyi]: Does it really need to be a list? The implementation doesn't seem to impose strong requirements, so `Sequence` would be better. This also applies to line 40 above.

File: stdlib/3/importlib/metadata.pyi [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+            def path(self) -> List[str]: ...
+            @property
+            def pattern(self) -> str: ...
+        @abc.abstractmethod
+        def find_distributions(self, context: Context = ...) -> Iterable[Distribution]: ...
+    class MetadataPathFinder(DistributionFinder):
+        @classmethod
+        def find_distributions(cls, context: DistributionFinder.Context = ...) -> Iterable[PathDistribution]: ...
+    class PathDistribution(Distribution):
+        def __init__(self, path: os.PathLike[str]) -> None: ...
```
Comment (JelleZijlstra) [inline stdlib/3/importlib/metadata.pyi]: This is too permissive:
```
        :param path: A pathlib.Path or similar object supporting
                     .joinpath(), __div__, .parent, and .read_text().
```
It needs to be either pathlib.Path or a custom protocol.

Comment (srittau) [reply stdlib/3/importlib/metadata.pyi]: I'm going for a `Path` for now for simplicity's sake. Can always be changed later to a custom protocol.

**Pull request merged**